### ISSUE TITLE
Runtime-defined map range doesn't work

### ISSUE THREAD

**Issue by am-ivanov** [spcl/dace#34]
**Describe the bug**
I am trying to use the input variable (not symbolic value) to define range of map iterations. It doesn't work (probably due to incorrect memlet propagation).

**To Reproduce**
```python
import dace
import numpy as np

N = dace.symbol('N')

@dace.program
def plus_1(X_in: dace.float32[N], num: dace.float32[1], X_out: dace.float32[N]):
    @dace.map
    def p1(i : _[0:num[0]]):
        x_in << X_in[i]
        x_out >> X_out[i]

        x_out = x_in + 1

X = np.random.rand(10).astype(np.float32)
Y = np.zeros(10)
num = np.zeros(1)
num[0] = 7

plus_1(X_in=X, num=num, X_out=Y, N=10)

print(Y)
```

It gives an error: `KeyError: 'Missing program argument "__p1_e0"'`

**Expected behavior**
First 7 elements of Y filled by non-zero random values from X with added 1 to them.

**Screenshots**
![image](https://user-images.githubusercontent.com/9387186/68898727-4e32d400-0730-11ea-98a3-f23f5a16e552.png)

![image](https://user-images.githubusercontent.com/9387186/68899391-d665a900-0731-11ea-83df-fdcbc66481e7.png)

(another problem: can't show all text on memlets simultaneously, this is why there are two screenshots)

**Comment by alexnick83** [comment]
This is not a bug. Data used in map ranges must be scalar. It would be nice if we could support also array elements, but we have to discuss with @tbennun how easy that would be. As a temporary measure, I will check if it is easy to detect this in the frontend and emit an appropriate error.

**Comment by am-ivanov** [comment]
Aren't single-element arrays (num: dace.float32[1]) scalars? How this code should be rewritten to have a correct behavior?

**Comment by alexnick83** [comment]
> Aren't single-element arrays (num: dace.float32[1]) scalars? How this code should be rewritten to have a correct behavior? @alexnick83

It has to be `dace.data.Scalar` (not an array, even if it has size 1). A way to write this would be:
```python
@dace.program
def plus_1(X_in: dace.float32[N], num: dace.int32, X_out: dace.float32[N]):
    @dace.map
    def p1(i : _[0:num]):
        x_in << X_in[i]
        x_out >> X_out[i]

        x_out = x_in + 1

X = np.random.rand(10).astype(np.float32)
Y = np.zeros(10, dtype=np.float32)
num = np.zeros(1, dtype=np.int32)
num[0] = 7

plus_1(X_in=X, num=num[0], X_out=Y, N=10)
```

You also have to be careful with the types. Default type for a numpy array is python `float`, which is `np.float64`. Python `int` is `np.int64`.
```

**Comment by tbennun** [comment]
Fixed in #35 .
